### PR TITLE
Add support for setting SM2 id field

### PR DESCRIPTION
**Pull request by randombit**
Fixing #6534 Documentation has not been added and tests are incomplete (need to plumb this into the EVP tests and at least verify a known signature with UID field set). Posting first for initial review of the approach.

- [ ] documentation is added or updated
- [ ] tests are added or updated

This is a contribution from Ribose Inc (@riboseinc)

### CODE REVIEW

File: crypto/sm2/sm2_pmeth.c [hunk 1]
```diff
@@ -149,6 +159,26 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
     EC_GROUP *group;
 
     switch (type) {
+    case EVP_PKEY_CTRL_DIGESTINIT:
```
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/sm2/sm2_pmeth.c]: I think it is not necessary, just a leftover from trying different approaches. Will remove.

File: crypto/evp/digest.c [hunk 2]
```diff
@@ -140,7 +140,22 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
...
+    if (ctx->digest->init(ctx) == 0)
+        return 0;
+
+    if (ctx->pctx) {
+        int r;
+        r = EVP_PKEY_CTX_ctrl(ctx->pctx, -1, EVP_PKEY_OP_TYPE_SIG,
+                              EVP_PKEY_CTRL_DIGESTSETUP, 0, ctx);
+
+        if (r == -2) {
+            ERR_clear_error();
```
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/evp/digest.c]: The problem is `EVP_PKEY_CTX_ctrl` does this:

```
    if (ret == -2)
        EVPerr(EVP_F_EVP_PKEY_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);

    return ret;
```

And here in digest.c we are swallowing the error return, since only SM2 uses this DIGESTSETUP PKEY flag. If we do not also clear the error state, it causes code downstream to fail. Specifically I noticed failures in the TLS layer tests unless I cleared the error state here, presumably because somewhere TLS looks at the error stack in order decide if something failed.

File: crypto/sm2/sm2_pmeth.c [hunk 3]
```diff
@@ -149,6 +159,26 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
...
+
+       if (dctx->uid != NULL) {
+           EC_KEY *ec = ctx->pkey->pkey.ec;
+           EVP_MD_CTX *md_ctx = (EVP_MD_CTX*)p2;
+           const EVP_MD* md = EVP_MD_CTX_md(md_ctx);
+           const int md_len = EVP_MD_size(md);
+           int rc = 1;
+           uint8_t *za = NULL;
+
+           za = OPENSSL_malloc(md_len);
```
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/sm2/sm2_pmeth.c]: Oh, yes that's much simpler. Will fix.

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -530,6 +532,12 @@ static int test_EVP_SM2(void)
     if (!TEST_ptr(md_ctx_verify = EVP_MD_CTX_new()))
         goto done;
 
+    if (!TEST_true(EVP_DigestSignInit(md_ctx, &sign_ctx, EVP_sm3(), NULL, pkey)))
+        goto done;
+
+    if (!TEST_true(EVP_PKEY_CTX_set_sm2_uid(sign_ctx, "nobody@example.com") > 0))
```
[the remark this answers is not in the document]
Comment (randombit) [reply test/evp_extra_test.c]: Sure, I will add this along with a "real" KAT wired into `evp_test`.

File: include/openssl/ec.h [hunk 5]
```diff
@@ -1342,6 +1342,10 @@ void EC_KEY_METHOD_get_verify(const EC_KEY_METHOD *meth,
                                 EVP_PKEY_OP_DERIVE, \
                                 EVP_PKEY_CTRL_GET_EC_KDF_UKM, 0, (void *)(p))
 
+# define EVP_PKEY_CTX_set_sm2_uid(ctx, uid) \
```
[the remark this answers is not in the document]
Comment (randombit) [reply include/openssl/ec.h]: Since the UID is tied to a particular `PKEY_CTX` I assumed the application already knows what value it set, but I suppose symmetric does argue for a getter.

File: crypto/sm2/sm2_pmeth.c [hunk 6]
```diff
@@ -167,6 +197,14 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
         EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
         return 1;
 
+    case EVP_PKEY_CTRL_SM2_SET_UID:
+        dctx->uid = OPENSSL_strdup(p2);
+        return 1;
+
+    case EVP_PKEY_CTRL_SM2_GET_UID:
+        *(char**)p2 = OPENSSL_strdup(dctx->uid);
```
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/sm2/sm2_pmeth.c]: Do you mean "is it really true that SM2 UIDs will always be null terminated strings"? I believe this is the case. But perhaps @ronaldtse knows more.

File: test/evp_extra_test.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (randombit) [reply test/evp_extra_test.c]: Oh well it turned out to be difficult (impossible?) to test a SM2 signature via `evp_test` because it expects all of the hash values to be provided in precomputed form, so instead I added another test to `evp_extra_test` which verifies a known signature created using the personalization field.

File: include/openssl/ec.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (randombit) [reply include/openssl/ec.h]: Done

**Comment by randombit** [comment]
Review comments should be addressed. There is also a proper test now.

File: crypto/evp/digest.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/evp/digest.c]: Perhaps, `ERR_set_mark()`, `ERR_pop_to_mark()`, `ERR_clear_last_mark()` would be better?

File: crypto/sm2/sm2_pmeth.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/sm2/sm2_pmeth.c]: I meant do we need to create a copy of it for the getter? Can't we just return a direct reference to `dctx->uid` (i.e. in OpenSSL terms we treat this as a `get0` instead of a `get1`)

File: crypto/evp/digest.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/evp/digest.c]: Seems so since that avoids wiping out the entire error stack. Thanks, didn't know about these.

File: include/openssl/ec.h [hunk 7]
```diff
@@ -1342,6 +1342,14 @@ void EC_KEY_METHOD_get_verify(const EC_KEY_METHOD *meth,
...
+        EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_SM2, EVP_PKEY_OP_TYPE_SIG, \
+                                EVP_PKEY_CTRL_SM2_SET_UID, 0, (void*)(uid))
+
+# define EVP_PKEY_CTX_get_sm2_uid(ctx, uid) \
```
[the remark this answers is not in the document]
Comment (randombit) [reply include/openssl/ec.h]: I will remove the strdup if that's preferable. Personally I prefer to avoid giving out internal pointers that might change over time, as it is easy to create dangling pointers that way if the application hangs on to the result longer than it should.

File: crypto/sm2/sm2_pmeth.c [hunk 8]
```diff
@@ -149,6 +159,22 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
...
+
+       if (dctx->uid != NULL) {
+           EC_KEY *ec = ctx->pkey->pkey.ec;
+           EVP_MD_CTX *md_ctx = (EVP_MD_CTX*)p2;
+           const EVP_MD* md = EVP_MD_CTX_md(md_ctx);
+           const int md_len = EVP_MD_size(md);
+           int rc = 1;
+           uint8_t za[EVP_MAX_MD_SIZE];
+
+           sm2_compute_userid_digest(za, md, dctx->uid, ec);
```
[the remark this answers is not in the document]
Comment (randombit) [reply crypto/sm2/sm2_pmeth.c]: Fixed

File: crypto/sm2/sm2_pmeth.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (ronaldtse) [reply crypto/sm2/sm2_pmeth.c]: @randombit SM2 UIDs (in DSA, PKE) makes no assumption to NULL-termination, the length of the UID is explicitly coded in `ENTL`. Perhaps there's no need to make a copy as suggested by @mattcaswell ?

File: include/openssl/ec.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (randombit) [reply include/openssl/ec.h]: strdup removed

**Comment by mattcaswell** [comment]
Been away...just trying to catch up. I'll try and take a look at this on Monday.

File: crypto/evp/digest.c [hunk 9]
```diff
@@ -140,7 +140,26 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
...
+
+        ERR_set_mark();
+
+        r = EVP_PKEY_CTX_ctrl(ctx->pctx, -1, EVP_PKEY_OP_TYPE_SIG,
+                              EVP_PKEY_CTRL_DIGESTSETUP, 0, ctx);
+
+        if (r == -2)
+            ERR_pop_to_mark();
+
+        ERR_clear_last_mark();
```
Comment (mattcaswell) [inline crypto/evp/digest.c]: `ERR_pop_to_mark()` also clears the last mark, so this call should be part of an `else` in the above `if`.

File: include/openssl/ec.h [hunk 10]
```diff
@@ -1342,6 +1342,14 @@ void EC_KEY_METHOD_get_verify(const EC_KEY_METHOD *meth,
                                 EVP_PKEY_OP_DERIVE, \
                                 EVP_PKEY_CTRL_GET_EC_KDF_UKM, 0, (void *)(p))
 
+# define EVP_PKEY_CTX_set1_sm2_uid(ctx, uid) \
+        EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_SM2, EVP_PKEY_OP_TYPE_SIG, \
+                                EVP_PKEY_CTRL_SM2_SET_UID, 0, (void*)(uid))
+
+# define EVP_PKEY_CTX_get1_sm2_uid(ctx, uid) \
```
Comment (mattcaswell) [inline include/openssl/ec.h]: We need to document these two macros too. You should mention that we expect a NUL terminated string for the uid. Also the memory ownership semantics.

File: test/evp_extra_test.c [hunk 11]
```diff
@@ -479,25 +480,100 @@ static int test_EVP_PKCS82PKEY(void)
...
+    if (!TEST_true(EVP_PKEY_set_alias_type(pkey, EVP_PKEY_SM2)))
+        goto done;
+
+    if (!TEST_ptr(md_ctx_verify = EVP_MD_CTX_new()))
+        goto done;
+
+    if (!TEST_true(EVP_DigestVerifyInit(md_ctx_verify, &verify_ctx, EVP_sm3(), NULL, pkey)))
+        goto done;
+
+    if (!TEST_true(EVP_PKEY_CTX_set1_sm2_uid(verify_ctx, user_id) > 0))
```
Comment (mattcaswell) [inline test/evp_extra_test.c]: `TEST_int_gt` might be better to use than `TEST_true` here.

File: include/openssl/ec.h (hunk 10 above)
Comment (mattcaswell) [reply include/openssl/ec.h]: Either this needs to be get0 and don't use strdup in the implementation, or get1 and use strdup. At the moment its get1 without strdup.

get0 == Reference to the internal data is returned; caller must not free
get1 == A new reference is returned, caller must free
set0 == Ownership of the reference is passed
set1 == Ownership of the reference is not passed (a copy is taken, or reference count is increased)

**Comment by randombit** [comment]
Review comments addressed

**Comment by ronaldtse** [comment]
Hoping this can be merged soon for #6837 and certificate generation. Thanks!

File: doc/man3/EVP_PKEY_CTX_ctrl.pod [hunk 12]
```diff
@@ -160,6 +164,14 @@ For maximum compatibility the named curve form should be used. Note: the
 B<OPENSSL_EC_NAMED_CURVE> value was only added to OpenSSL 1.1.0; previous
 versions should use 0 instead.
 
+The EVP_PKEY_CTX_set1_sm2_uid() sets the user identifier string, which is used
+for the SM2 signature scheme. The uid string is assumed to be NULL terminated,
```
Comment (mattcaswell) [inline doc/man3/EVP_PKEY_CTX_ctrl.pod]: Nit: For strings we talk about NUL-terminated (i.e. single L).

**Comment by mattcaswell** [comment]
My reading is that the UID is fairly critical for SM2, so I guess this is a "bug" that it wasn't there before, and hence should be in 1.1.1.

**Comment by ronaldtse** [comment]
I agree this should be considered a bug. On the other hand, we still don’t have certificate generation so the functionality of SM2 isn’t yet complete.

**Comment by mattcaswell** [comment]
> On the other hand, we still don’t have certificate generation so the functionality of SM2 isn’t yet complete.

Right...but SM2 is useful by itself without certificates. So certificate support seems like a new feature.

**Comment by ronaldtse** [comment]
Agreed! :+1

File: crypto/sm2/sm2_pmeth.c [hunk 13]
```diff
@@ -149,6 +159,21 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
     EC_GROUP *group;
 
     switch (type) {
+    case EVP_PKEY_CTRL_DIGESTSETUP:
+
+       if (dctx->uid != NULL) {
+           EC_KEY *ec = ctx->pkey->pkey.ec;
+           EVP_MD_CTX *md_ctx = (EVP_MD_CTX*)p2;
+           const EVP_MD* md = EVP_MD_CTX_md(md_ctx);
```
Comment (dot-asm) [inline crypto/sm2/sm2_pmeth.c]: Styling guide stipulates that it should `(EVP_MD_CTX *)p1` and `EVP_MD *md`.

File: crypto/sm2/sm2_pmeth.c [hunk 14]
```diff
@@ -63,6 +66,13 @@ static int pkey_sm2_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
             return 0;
         }
     }
+
+    dctx->uid = OPENSSL_strdup(sctx->uid);
+    if (sctx->uid != NULL && dctx->uid == NULL) {
```
Comment (dot-asm) [inline crypto/sm2/sm2_pmeth.c]: Even though OPENSSL_strdup can handle NULL, I'd still argue in favour of same semantic as conventional strdup. I.e. check sctx->uid for NULL *first*. It's more readable that way. Or in other words for example
```
if (sctx->uid != NULL
    && (dctx->uid = OPENSSL_strdup(sctx->uid) == NULL)
```

File: crypto/sm2/sm2_pmeth.c [hunk 15]
```diff
@@ -167,6 +192,14 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
         EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
         return 1;
 
+    case EVP_PKEY_CTRL_SM2_SET_UID:
+        dctx->uid = OPENSSL_strdup(p2);
```
Comment (dot-asm) [inline crypto/sm2/sm2_pmeth.c]: What if dctx->uid wasn't NULL? Would it be appropriate to free it?

File: crypto/sm2/sm2_pmeth.c [hunk 16]
```diff
@@ -167,6 +192,14 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
...
+        return (dctx->uid != NULL);
+
+    case EVP_PKEY_CTRL_SM2_GET_UID:
+        *(char**)p2 = dctx->uid;
```
Comment (dot-asm) [inline crypto/sm2/sm2_pmeth.c]: `(char **)`

File: doc/man3/EVP_DigestSignInit.pod [hunk 17]
```diff
@@ -29,6 +29,13 @@ EVP_PKEY_CTX of the signing operation will be written to B<*pctx>: this can
 be used to set alternative signing options. The digest B<type> may be NULL if
 the signing algorithm supports it.
 
+In some rare cases (including SM2 signatures) it is necessary to reinitialize
+the digest computation after setting some parameters on the B<pctx>. This
+is done by calling EVP_DigestSignInit() twice. The first time, B<pctx>
```
Comment (dot-asm) [inline doc/man3/EVP_DigestSignInit.pod]: I'm no expert in English, but I find "twice" confusing. I give an impression that I actually have to call it two times in a row prior I can do anything.

File: include/openssl/ec.h [hunk 18]
```diff
@@ -1342,6 +1342,14 @@ void EC_KEY_METHOD_get_verify(const EC_KEY_METHOD *meth,
                                 EVP_PKEY_OP_DERIVE, \
                                 EVP_PKEY_CTRL_GET_EC_KDF_UKM, 0, (void *)(p))
 
+# define EVP_PKEY_CTX_set1_sm2_uid(ctx, uid) \
+        EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_SM2, EVP_PKEY_OP_TYPE_SIG, \
+                                EVP_PKEY_CTRL_SM2_SET_UID, 0, (void*)(uid))
+
+# define EVP_PKEY_CTX_get0_sm2_uid(ctx, uid) \
+        EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_SM2, EVP_PKEY_OP_TYPE_SIG, \
+                                EVP_PKEY_CTRL_SM2_GET_UID, 0, (void*)(uid))
```
Comment (dot-asm) [inline include/openssl/ec.h]: Formatting nits. Two lines are one space short. Well, they are formatted after previous defines, but it doesn't mean that previous defines are correctly formatted, and should not be reason for perpetuating incorrect formatting.

File: test/evp_extra_test.c [hunk 19]
```diff
@@ -479,25 +480,100 @@ static int test_EVP_PKCS82PKEY(void)
...
+       "AgEBA0IABArkx3mKoPEZRxvuEYJb5GICu3nipYRElel8BP9N8lSKfAJA+I8c1OFj\n"
+       "Uqc8F7fxbwc1PlOhdtaEqf4Ma7eY6Fc=\n"
+       "-----END PUBLIC KEY-----\n";
+
+    const char *input = "message digest";
+    const char *user_id = "ALICE123@YAHOO.COM";
+
+    const uint8_t signature[] = {
+       0x30, 0x44, 0x02, 0x20,
+       0x40, 0xF1, 0xEC, 0x59, 0xF7, 0x93, 0xD9, 0xF4, 0x9E, 0x09, 0xDC, 0xEF, 0x49, 0x13, 0x0D, 0x41, 0x94, 0xF7, 0x9F, 0xB1, 0xEE, 0xD2, 0xCA, 0xA5, 0x5B, 0xAC, 0xDB, 0x49, 0xC4, 0xE7, 0x55, 0xD1,
```
Comment (dot-asm) [inline test/evp_extra_test.c]: Line length limit.

File: doc/man3/EVP_DigestSignInit.pod (hunk 17 above)
Comment (dot-asm) [reply doc/man3/EVP_DigestSignInit.pod]: I can see the "twice" in test and ... it's terrible...

Comment (randombit) [reply doc/man3/EVP_DigestSignInit.pod]: I agree it is, uh, well terrible is a pretty good word for it tbh. It was suggested by mattcaswell in #6534 and I couldn't come up with an alternative that would be acceptable.

**Comment by randombit** [comment]
Review comments addressed (modulo the non-optimal API for setting the UID, but it's not clear how else to make this work).

File: doc/man3/EVP_DigestSignInit.pod (hunk 17 above)
Comment (mattcaswell) [reply doc/man3/EVP_DigestSignInit.pod]: Yes. It's quite nasty. The other alternative I suggested in #6534 was to create a new version of `EVP_DigestSignInit`. But I'm not especially keen on that idea either (it seems like a "sledgehammer to crack a nut"). IMO the original approach in #6534 was worse, so I'm out of ideas. This seemed like the least bad option.

Comment (ronaldtse) [reply doc/man3/EVP_DigestSignInit.pod]: @dot-asm would it make it better if we reword this to clarify? Such as:

"
The SM2 algorithm accepts a personalization string, if you do not need to customize the personalization string, you can use `EVP_DigestSignInit()` as expected (just call once and use it). 

However, if you need to customize the personalization string, such as processing of SM2 signatures, it is necessary to reinitialize the digest computation after setting some parameters on the `B<pctx>` by calling `EVP_DigestSignInit()` *twice*. The rationale for this is described in https://github.com/openssl/openssl/issues/6534.

* First, call EVP_DigestSignInit() to get a handle on the `EVP_PKEY_CTX`. The first time, `B<pctx>` should be non `NULL`.
* Then, set the necessary options on `B<pctx>`, for instance, using `EVP_PKEY_CTX_ctrl_str()`.
* Afterwards, `EVP_SignVerifyInit()` is called again, using the same verification context `B<ctx>`.

"

Comment (dot-asm) [reply doc/man3/EVP_DigestSignInit.pod]: > the least bad option.

Which in turn is like worst excuse. As already said [much] earlier, if problem can't be solved in reasonable manner, we would be better off not solving it in minor release. Because introduction of kludges obliges us to maintain them. And they are likely to be dead end, in sense that in major release problem is likely to be solved differently. It's really loose-loose...

However. As far as I can tell, if EVP_DigestVerifyUpdate involved pmeth in performing updates, as opposite to be just an alias to EVP_DigestUpdate, single call would suffice. Right? I mean pmeth could catch first update and inject additional [personalization] data. It should be noted that EVP_MD_CTX has "virtual" update function. ("Virtual" is reference to C++-style virtual methods). This mean that pmeth can get involved in performing updates by overriding update function in EVP_MD_CTX. In this case it can even restore original function after first call, so that it wouldn't have to be involved past first invocation...

On related note I'd like to argue against naming new ctrl as sm2_uid. In sense that I'd prefer more generic name that could be reusable. For example why not personalization?

**Comment by dot-asm** [comment]
I'm not quite clear what is target here. Last said words were about personalization data being a feature, yet "post 1.1.1" milestone was changed to "1.1.1"...

**Comment by mattcaswell** [comment]
> Last said words were about personalization data being a feature, yet "post 1.1.1" milestone was changed to "1.1.1"...

No, we agreed above that UID is required for SM2. Hence the 1.1.1 milestone. The discussion above was that *certificate* support was a feature (which also depends on the UID support).

**Comment by dot-asm** [comment]
Well, I disagree with this assessment. It's more of a feature than some of others that were disqualified. At least it's clearly visible in public headers...

**Comment by mattcaswell** [comment]
> Well, I disagree with this assessment. It's more of a feature than some of others that were disqualified. At least it's clearly visible in public headers...

My understanding is that we can't really use SM2 without it because there is required to be a UID present.

**Comment by dot-asm** [comment]
Common "mistake" seems to be that it's assumed that it's sufficient to add an algorithm and then magic would happen by itself. But it never does. Focus should really be on how implementation interacts with outer world. For example you say that "we can't really use SM2 without it because there is required to be a UID present." But it's still about algorithm implementation. Customarily signature is structured data blob that describes algorithms and parameters, and question should be how *that* is supposed to work, not how to **hammer** stuff into box and make it crack, so that one can write tailored test program. In other words even with this it's not as useful as user would expect.

**Comment by mattcaswell** [comment]
> Common "mistake" seems to be that it's assumed that it's sufficient to add an algorithm and then magic would happen by itself. But it never does. Focus should really be on how implementation interacts with outer world. For example you say that "we can't really use SM2 without it because there is required to be a UID present." But it's still about algorithm implementation. Customarily signature is structured data blob that describes algorithms and parameters, and question should be how that is supposed to work, not how to hammer stuff into box and make it crack, so that one can write tailored test program. In other words even with this it's not as useful as user would expect.

I don't really understand your objection or what you are trying to say here. Are you suggesting we should release SM2 without UID support? That seems to be the default we are going to end up with unless we can resolve this.

**Comment by dot-asm** [comment]
> I don't really understand your objection or what you are trying to say here.

What I'm trying to say is that it doesn't work in long run to practice this way. That is to just hammer stuff in till *something* implementation-specific works. Initial SM2 support that was hammered into ecdsa_ossl.c? Not the way to do it, was resolved. Making this work through double call? I'd say as bad, not the way to do it. And we choice is between hammering and releasing without something, I'd say latter. This is what I'm trying to say. As for objections, there is "however" paragraph in https://github.com/openssl/openssl/pull/6757#discussion_r208184883. Though I'm not saying that it would do, only that it *might*. As already said, one should view it from application side, i.e. how it would be used in implementation-neutral way, rather than from algorithm side, i.e. implement algorithm and what it takes to use it is application problem.

File: doc/man3/EVP_DigestSignInit.pod [hunk 20]
```diff
@@ -29,6 +29,13 @@ EVP_PKEY_CTX of the signing operation will be written to B<*pctx>: this can
...
+should be non NULL. Then any necesssary options can be set on B<pctx>
+using for instance EVP_PKEY_CTX_ctrl_str(). Afterwards, EVP_SignVerifyInit()
```
Comment (InfoHunter) [inline doc/man3/EVP_DigestSignInit.pod]: Doc nit: it seems this should be: "Afterwards, EVP_DigestSignInit()...."

**Comment by InfoHunter** [comment]
Since SM2 *requires* an ID/UID string when forming the 'prefix'(`ZA`) of the being signed message, the current implementation is not correct (more accurate, it has a bug...). To fix this bug, I would like to agree with @mattcaswell , or there might be a better way on the implementation side?

Furthermore, as discussed in relevant PR #6837 , we should be able to use the `SM2_DEFAULT_USERID` when forming the prefix `ZA` in SM2 sign/verify procedure if the user doesn't set a UID. Thus I guess if it's possible, then PR #6837 seems avoidable...

**Comment by InfoHunter** [comment]
Ping @randombit , what's your opinion on this?

**Comment by randombit** [comment]
@InfoHunter Using the default user ID for SM2 signatures in any situation where the application has not set it probably makes sense. @ronaldtse what do you think?

**Comment by ronaldtse** [comment]
> Using the default user ID for SM2 signatures in any situation where the application has not set it probably makes sense.

I fully agree.

**Comment by t-j-h** [comment]
-1 blocker for now until we get through this context as it isn't making sense to have a "default" for something that is always meant to be set by the user. 

The concept that there are test cases that need this set doesn't remove the idea that without it being set there should be an error.

I think we need to look at the usage here before we hardwire in something like this.

**Comment by ronaldtse** [comment]
@t-j-h the default user ID is specifically demanded by GM/T 0009 Clause 10. GM/T 0009 is the document that specifies usage of the SM2 algorithm:

https://riboseinc.github.io/gmt-0009-2012/

The User ID field is only meant to be customized by the user with pre-agreement.

**Comment by t-j-h** [comment]
I think you are mixing things up here.
ISO/IEC 14888-3:2016 does not define that there is a default "identity string" at all and that is the low level algorithm document.
If there are other standards that are layered on top which specify what to do if an "identity string" is not specified then those (to me at least) are meant to be handled elsewhere.

**Comment by t-j-h** [comment]
The comment from @dot-asm on the name for this function and ID that does not contain SM2 should be addressed. ISO/IEC 14888-3:2016 calls it "identity string" and we could call it "personalisation string" for a generic term as suggested by others - but SM2_UID is not the right naming. Please fix that.

**Comment by ronaldtse** [comment]
@t-j-h just for the record, both GM/T 0003 and GB/T 32918 (Parts 2-4), which are the official specifications of SM2, require usage of the default UID string in its tests. ISO/IEC 14888-3 "adopts" SM2 from these two documents.

In particular, the GM/T 0009 and GM/T 0010 documents are normative references of GB/T 32918, and defers to them within the standard. In a Standards point of view, both 0009 and 0010 form part of the official SM2 standard.

**Comment by t-j-h** [comment]
@ronaldtse I have an official copy of ISO/IEC 14888-3 and have read it in detail. I have read translations of the GM/T docs too. 

The alternative approach in #7113 allows for the user setting the identity string and is layered in a manner which allows for multiple usage contexts and nothing precludes the outer code from using the functionality in a manner which passes the tests.

There is no requirement for a default value in the ISO specification or in the base SM2 specification document either, and it is also made clear that the value itself is an OCTET STRING which may have any valid length (including 0) conceptually - it is not an ASCII string. The base primitive as published does not wire in a default and the documents that suggest the default value are themselves rather terse and incomplete and do not override the base definition. 

This particular PR contained a number of assumptions which were incorrect and certainly not covered in the base specification. The alternate PR allows for the functionality this covers and does not restrict the usage model and also avoids the double calls to EVP_DigestSignInit and SM2-specific labeling of the concepts. 

The functionality in #7113 allows OpenSSL users to perform all the required capability without the limitations and assumptions that were in this PR. There isn't anything which is missing in terms of being able to drive the underlying crypto in order to pass the tests. it is a different way to solve the problem and it fits into the code base more cleanly and avoids SM2-specific items an invalid assumptions about ASCII-only. This is why we have a review process in place and why we need PRs like this one put forward for review so that multiple people can look at it and evaluate a range of approaches. Your contributions in this PR are valued and welcome and that is how collaboration works so we get better outcomes. 

@InfoHunter has addressed all of the issues noted above and also your PR and his have encouraged more people to read the specification and to look into how to best fit this capability within OpenSSL.

**Comment by ronaldtse** [comment]
@t-j-h , I agree that @InfoHunter 's approach is more suitable for OpenSSL and we have no objections to the newer PR.

> the documents that suggest the default value are themselves rather terse and incomplete

I suppose this comment is for GM/T 0009 and GM/T 0010. They are regardless still "base" normative documents to a whole body of work relating to SM2, including certificate formats. Glad I wasn't involved in these. 😉
